### PR TITLE
Refactors to generation in preparation for compose

### PR DESCRIPTION
**Pull request by smarterclayton**
Should have 1:1 output, but makes it easier to do deeper generation.

@csrwng @bparees (may still have broken tests)

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test]

File: pkg/generate/app/sourcelookup.go [hunk 1]
```diff
@@ -116,6 +116,27 @@ func NewSourceRepository(s string) (*SourceRepository, error) {
 	}, nil
 }
 
+// NewSourceRepository creates a reference to a local or remote source code repository from
+// a URL or path.
```
Comment (bparees) [inline pkg/generate/app/sourcelookup.go]: NewSourceRepositoryWithDockerfile

File: pkg/generate/app/imageref.go [hunk 2]
```diff
@@ -143,15 +143,20 @@ type ImageRef struct {
 	// If true, create the image stream using a tag for this reference, not a bulk
 	// import.
 	TagDirectly bool
+	// If set, the tag that othre components will reference this image by. Must be
```
Comment (bparees) [inline pkg/generate/app/imageref.go]: s/othre/other/

**Comment by smarterclayton** [comment]
Any other comments?

**Comment by bparees** [comment]
not from me.

File: pkg/generate/app/app.go [hunk 3]
```diff
@@ -432,3 +435,62 @@ func LabelsFromSpec(spec []string) (map[string]string, []string, error) {
 	}
 	return labels, remove, nil
 }
+
+// TODO: move to pkg/runtime or pkg/api
+func AsVersionedObjects(objects []runtime.Object, typer runtime.ObjectTyper, convertor runtime.ObjectConvertor, versions ...unversioned.GroupVersion) []error {
```
Comment (csrwng) [inline pkg/generate/app/app.go]: where is this used?

File: pkg/generate/app/imageref.go [hunk 4]
```diff
@@ -143,15 +143,20 @@ type ImageRef struct {
 	// If true, create the image stream using a tag for this reference, not a bulk
 	// import.
 	TagDirectly bool
+	// If set, the tag that other components will reference this image by. Must be
```
Comment (csrwng) [inline pkg/generate/app/imageref.go]: Use godoc convention

File: pkg/generate/app/pipeline.go [hunk 5]
```diff
@@ -296,12 +273,10 @@ func makeValidServiceName(name string) (string, string) {
...
 type sortablePorts []kapi.ContainerPort
 
-func (s sortablePorts) Len() int           { return len(s) }
-func (s sortablePorts) Less(i, j int) bool { return s[i].ContainerPort < s[j].ContainerPort }
-func (s sortablePorts) Swap(i, j int) {
-	p := s[i]
-	s[i] = s[j]
-	s[j] = p
+func (s sortablePorts) Len() int      { return len(s) }
+func (s sortablePorts) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
```
Comment (csrwng) [inline pkg/generate/app/pipeline.go]: interesting... hadn't seen this before. Nice to know that it works :)

File: pkg/generate/app/app.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/generate/app/app.go]: In the lib compose followup.

**Comment by smarterclayton** [comment]
Any more comments?

**Comment by bparees** [comment]
nothing other than you should merge this first so i don't have to rebase:
https://github.com/openshift/origin/pull/8323

:)  (kidding)

**Comment by csrwng** [comment]
Nope, LGTM

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Pull request merged**